File: src/SimpleCRM/Helpers/ObservableCollectionExtensions.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Linq;
using System.Reflection;

namespace SimpleCRM
{
    public static class ObservableCollectionExtensions
    {
        public static void AddRange<T>(this ObservableCollection<T> collection, IEnumerable<T> items)
        {
            foreach (var item in items)
            {
                collection.Add(item);
            }
        }

        public static void InsertRange<T>(this ObservableCollection<T> collection, IEnumerable<T> items)
        {
            var enumerable = items as List<T> ?? items.ToList();
            if (collection == null || items == null || !enumerable.Any())
            {
                return;
            }

            Type type = collection.GetType();

            type.InvokeMember("CheckReentrancy", BindingFlags.Instance | BindingFlags.InvokeMethod | BindingFlags.NonPublic, null, collection, null);
            var itemsProp = type.BaseType.GetProperty("Items", BindingFlags.NonPublic | BindingFlags.FlattenHierarchy | BindingFlags.Instance);
            var privateItems = itemsProp.GetValue(collection) as IList<T>;
            foreach (var item in enumerable)
            {
                privateItems.Add(item);
            }

            type.InvokeMember("OnPropertyChanged", BindingFlags.Instance | BindingFlags.InvokeMethod | BindingFlags.NonPublic, null,
              collection, new object[] { new PropertyChangedEventArgs("Count") });

            type.InvokeMember("OnPropertyChanged", BindingFlags.Instance | BindingFlags.InvokeMethod | BindingFlags.NonPublic, null,
              collection, new object[] { new PropertyChangedEventArgs("Item[]") });

            type.InvokeMember("OnCollectionChanged", BindingFlags.Instance | BindingFlags.InvokeMethod | BindingFlags.NonPublic, null,
              collection, new object[] { new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset) });
        }
    }
}

File: src/SampleCRM/Helpers/ObservableCollectionExtensions.vb
Imports System.Runtime.CompilerServices
Imports System.Collections.ObjectModel
Imports System.Collections.Specialized
Imports System.ComponentModel
Imports System.Reflection

Module ObservableCollectionExtensions
    <Extension()>
    Sub AddRange(Of T)(ByVal collection As ObservableCollection(Of T), ByVal items As IEnumerable(Of T))
        For Each item In items
            collection.Add(item)
        Next
    End Sub

    <Extension()>
    Sub InsertRange(Of T)(ByVal collection As ObservableCollection(Of T), ByVal items As IEnumerable(Of T))
        Dim enumerable = If(TryCast(items, List(Of T)), items.ToList())

        If collection Is Nothing OrElse items Is Nothing OrElse Not enumerable.Any() Then
            Return
        End If

        Dim type As Type = collection.[GetType]()
        type.InvokeMember("CheckReentrancy", BindingFlags.Instance Or BindingFlags.InvokeMethod Or BindingFlags.NonPublic, Nothing, collection, Nothing)
        Dim itemsProp = type.BaseType.GetProperty("Items", BindingFlags.NonPublic Or BindingFlags.FlattenHierarchy Or BindingFlags.Instance)
        Dim privateItems = TryCast(itemsProp.GetValue(collection), IList(Of T))

        For Each item In enumerable
            privateItems.Add(item)
        Next

        type.InvokeMember("OnPropertyChanged", BindingFlags.Instance Or BindingFlags.InvokeMethod Or BindingFlags.NonPublic, Nothing, collection, New Object() {New PropertyChangedEventArgs("Count")})
        type.InvokeMember("OnPropertyChanged", BindingFlags.Instance Or BindingFlags.InvokeMethod Or BindingFlags.NonPublic, Nothing, collection, New Object() {New PropertyChangedEventArgs("Item[]")})
        type.InvokeMember("OnCollectionChanged", BindingFlags.Instance Or BindingFlags.InvokeMethod Or BindingFlags.NonPublic, Nothing, collection, New Object() {New NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset)})
    End Sub
End Module
